matching/src/KorePatterns.hs

```diff
@@ -1,51 +1,74 @@
-{-# LANGUAGE GADTs           #-}
-{-# LANGUAGE PatternSynonyms #-}
-
 module KorePatterns where
 
+import Control.Comonad.Trans.Cofree
+       ( CofreeF (..) )
+import Data.Functor.Foldable as Recursive
+
+
 import Kore.AST.Common
        ( And (..), Application (..), Bottom (..), Equals (..), Implies (..),
-       Pattern (..), Rewrites (..), Sort (..), SymbolOrAlias (..), Top (..) )
+       Pattern (..), Rewrites (..), SymbolOrAlias (..), Top (..) )
 import Kore.AST.Kore
-       ( pattern KoreObjectPattern, KorePattern )
+       ( KorePattern, UnifiedPattern(..) )
 import Kore.AST.MetaOrObject
-       ( Object (..) )
-
-pattern Implies_ :: Sort Object    -- ^ The sort level is fixed for our patterns
-                 -> KorePattern v  -- ^ The hypothesis
-                 -> KorePattern v  -- ^ The conclusion
-                 -> KorePattern v  -- ^ The resulting pattern
-pattern Implies_ sort hyp cnc = KoreObjectPattern (ImpliesPattern (Implies sort hyp cnc))
-
-pattern App_ :: SymbolOrAlias Object -- ^ The symbol that points to the function that is being invoked
-             -> [KorePattern v]      -- ^ The arguments passed to the function
-             -> KorePattern v        -- ^ The resulting pattern
-pattern App_ fnSymbol args =
+       ( Object (..), Meta (..) )
+import Kore.Sort
+       ( Sort (..) )
+
+pattern KoreObjectPattern :: Functor domain
+                          => Pattern Object domain variable
+                              (KorePattern domain variable annotation)
+                          -> KorePattern domain variable annotation
+pattern KoreObjectPattern pat <- (Recursive.project -> _ :< UnifiedObjectPattern pat)
```

Comment:
A bit unfortunate that this is a unidirectional pattern, but I don't see any good way to invent an annotation.